pkg/scheduler/scheduler.go

```diff
@@ -612,50 +612,50 @@ func updateAssignmentForTAS(snapshot *schdcache.Snapshot, cq *schdcache.ClusterQ
 // assuming it in the cache.
 func (s *Scheduler) admit(ctx context.Context, e *entry, cq *schdcache.ClusterQueueSnapshot) error {
 	log := ctrl.LoggerFrom(ctx)
-	newWorkload := e.Obj.DeepCopy()
-	origWorkload := e.Obj.DeepCopy()
 	admission := &kueue.Admission{
 		ClusterQueue:      e.ClusterQueue,
 		PodSetAssignments: e.assignment.ToAPI(),
 	}
 
-	workload.SetQuotaReservation(newWorkload, admission, s.clock)
-	if workload.HasAllChecks(newWorkload, workload.AdmissionChecksForWorkload(log, newWorkload, cq.AdmissionChecks, schdcache.AllFlavors(cq.ResourceGroups))) {
-		// sync Admitted, ignore the result since an API update is always done.
-		_ = workload.SyncAdmittedCondition(newWorkload, s.clock.Now())
-	}
-	if err := s.cache.AssumeWorkload(log, newWorkload); err != nil {
-		return err
-	}
-	e.status = assumed
-	log.V(2).Info("Workload assumed in the cache")
+	wl := e.Obj.DeepCopy()
 
-	if afs.Enabled(s.admissionFairSharing) {
-		s.updateEntryPenalty(log, e, add)
+	err := workload.PatchAdmissionStatus(ctx, s.client, wl, s.clock, func() (bool, error) {
+		workload.SetQuotaReservation(wl, admission, s.clock)
+		if workload.HasAllChecks(wl, workload.AdmissionChecksForWorkload(log, wl, cq.AdmissionChecks, schdcache.AllFlavors(cq.ResourceGroups))) {
+			// sync Admitted, ignore the result since an API update is always done.
+			_ = workload.SyncAdmittedCondition(wl, s.clock.Now())
+		}
+		if err := s.cache.AssumeWorkload(log, wl); err != nil {
+			return false, err
+		}
+		e.status = assumed
+		log.V(2).Info("Workload assumed in the cache")
 
-		// Trigger LocalQueue reconciler to apply any pending penalties
-		s.queues.NotifyWorkloadUpdateWatchers(e.Obj, newWorkload)
-	}
+		if afs.Enabled(s.admissionFairSharing) {
+			s.updateEntryPenalty(log, e, add)
 
-	if features.Enabled(features.TopologyAwareScheduling) && workload.HasUnhealthyNodes(e.Obj) {
-		log.V(5).Info("Clearing the topology assignment recovery field from the workload status after successful recovery")
-		newWorkload.Status.UnhealthyNodes = nil
+			// Trigger LocalQueue reconciler to apply any pending penalties
+			s.queues.NotifyWorkloadUpdateWatchers(e.Obj, wl)
+		}
+
+		if features.Enabled(features.TopologyAwareScheduling) && workload.HasUnhealthyNodes(e.Obj) {
+			log.V(5).Info("Clearing the topology assignment recovery field from the workload status after successful recovery")
+			wl.Status.UnhealthyNodes = nil
+		}
+		return true, nil
+	}, workload.WithLooseOnApply())
+	if err == nil {
+		// Record metrics and events for quota reservation and admission
+		s.recordWorkloadAdmissionMetrics(wl, e.Obj, admission)
+
+		log.V(2).Info("Workload successfully admitted and assigned flavors", "assignments", admission.PodSetAssignments)
+		return nil
 	}
 
 	s.admissionRoutineWrapper.Run(func() {
-		err := workload.PatchAdmissionStatus(ctx, s.client, origWorkload, s.clock, func() (*kueue.Workload, bool, error) {
```

Comment:
/hold
We should not be moving the patch outside of the goroutine for the performance reasons